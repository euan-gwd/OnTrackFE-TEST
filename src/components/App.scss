.app-container {
  margin: 1rem 3rem;
  display: grid;
  grid-template-rows: 4rem 2rem max-content 1fr max-content;
  grid-template-columns: 95vw;
  grid-gap: 1rem 0;

  .filters {
    display: flex;
    justify-content: flex-end;
  }
}

.app-header {
  text-align: center;
}

.app-spinner-container {
  margin: 1rem 3rem;
  height: 100vh;
  display: grid;

  .app-spinner {
    margin: auto;
  }
}
