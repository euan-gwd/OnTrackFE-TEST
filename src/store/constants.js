export const GET_BOOKS_BEGIN = 'GET_BOOKS_BEGIN';
export const GET_BOOKS_SUCCESS = 'GET_BOOKS_SUCCESS';
export const GET_BOOKS_ERROR = 'GET_BOOKS_ERROR';
export const CHANGE_PAGE = 'CHANGE_PAGE';
