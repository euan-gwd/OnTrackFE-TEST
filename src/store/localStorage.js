export const loadState = () => {
  try {
    const savedStateInLocalStorage = sessionStorage.getItem('state');
    if (savedStateInLocalStorage === null) {
      return undefined;
    }

    return JSON.parse(savedStateInLocalStorage);
  } catch (error) {
    console.error(error);
    return undefined;
  }
};

export const saveState = state => {
  try {
    const savedState = JSON.stringify(state);
    sessionStorage.setItem('state', savedState);
  } catch (error) {
    console.error(error);
  }
};
