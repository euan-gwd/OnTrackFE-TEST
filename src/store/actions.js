import axios from 'axios';
import * as actionTypes from './constants';

export function getBooks(pageOptions) {
  return (dispatch, getState) => {
    dispatch({ type: actionTypes.GET_BOOKS_BEGIN });
    return axios
      .post(`http://nyx.vima.ekt.gr:3000/api/books/`, pageOptions)
      .then(res => {
        dispatch({
          type: actionTypes.GET_BOOKS_SUCCESS,
          payload: {
            books: res.data.books,
            count: res.data.count,
            page: pageOptions.page,
            itemsPerPage: pageOptions.itemsPerPage
          }
        });
      })
      .catch(error =>
        dispatch({
          type: actionTypes.GET_BOOKS_ERROR,
          payload: {
            error
          }
        })
      );
  };
}
