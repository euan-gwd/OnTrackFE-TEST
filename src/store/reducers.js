import * as actionTypes from './constants';

const initialState = {
  page: 1,
  itemsPerPage: 20,
  isLoading: false,
  error: null,
  books: [],
  totalRecords: 0
};

export function bookReducer(state = initialState, action) {
  switch (action.type) {
    case actionTypes.GET_BOOKS_BEGIN:
      return {
        ...state,
        isLoading: true,
        error: null
      };
    case actionTypes.GET_BOOKS_SUCCESS:
      return {
        ...state,
        books: action.payload.books,
        totalRecords: action.payload.count,
        page: action.payload.page,
        itemsPerPage: action.payload.itemsPerPage,
        isLoading: false,
        error: null
      };
    case actionTypes.GET_BOOKS_ERROR:
      return {
        ...state,
        error: action.payload.error,
        books: [],
        isLoading: false
      };
    default:
      return state;
  }
}
