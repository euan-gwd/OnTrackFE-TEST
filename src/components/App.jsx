import React, { Component } from 'react';
import { connect } from 'react-redux';
import { getBooks } from '../store/actions';
import ListGroup from 'react-bootstrap/ListGroup';

import Alert from 'react-bootstrap/Alert';
import Loader from './Loader';
import SearchBar from './SearchBar';
import PageItemsFilter from './PageItemsFilter';
import BookItem from './BookItem';
import PageNavBar from './PageNavBar';
import './App.scss';

class App extends Component {
  componentDidMount = () => {
    const { fetchBooks, page, itemsPerPage } = this.props;
    const pageOptions = { page, itemsPerPage, filters: [] };
    fetchBooks(pageOptions);
  };

  render() {
    const { books, isLoading, error } = this.props;

    if (isLoading) {
      return <Loader />;
    }

    if (error) {
      return (
        <Alert variant="danger">
          <Alert.Heading>Oops, sorry things fell down</Alert.Heading>
          <p>{error.message}</p>
        </Alert>
      );
    }

    return (
      <div className="app-container">
        <h1 className="app-header">The Library</h1>

        <SearchBar />

        <PageItemsFilter />

        <ListGroup>
          {books.map(book => (
            <ListGroup.Item key={book.id}>
              <BookItem book={book} />
            </ListGroup.Item>
          ))}
        </ListGroup>

        <PageNavBar />
      </div>
    );
  }
}

const mapStateToProps = state => ({
  books: state.books,
  isLoading: state.isLoading,
  error: state.error,
  page: state.page,
  itemsPerPage: state.itemsPerPage,
  filters: state.filters
});

const mapDispatchToProps = (dispatch, pageOptions) => ({
  fetchBooks: pageOptions => dispatch(getBooks(pageOptions))
});

export default connect(mapStateToProps, mapDispatchToProps)(App);
