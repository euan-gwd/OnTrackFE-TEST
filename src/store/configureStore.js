import { createStore, applyMiddleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension/developmentOnly';
import { bookReducer } from './reducers';
import { loadState, saveState } from './localStorage';
import thunk from 'redux-thunk';
import throttle from 'lodash/throttle';

const configureStore = () => {
  const persistedState = loadState();

  const store = createStore(bookReducer, persistedState, composeWithDevTools(applyMiddleware(thunk)));

  store.subscribe(
    throttle(() => {
      saveState({
        page: store.getState().page,
        itemsPerPage: store.getState().itemsPerPage,
        books: store.getState().books,
        totalRecords: store.getState().totalRecords
      });
    }, 1000)
  );

  return store;
};

export default configureStore;
