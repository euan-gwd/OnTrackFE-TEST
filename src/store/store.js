import { createStore, applyMiddleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension/developmentOnly';
import { bookReducer } from './reducers';
import { loadState } from './localStorage';
import thunk from 'redux-thunk';

const persistedState = loadState();

const store = createStore(
  bookReducer,
  persistedState,
  composeWithDevTools(applyMiddleware(thunk))
);

export default store;
