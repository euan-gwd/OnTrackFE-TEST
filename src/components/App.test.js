import React from 'react';
import { shallow } from 'enzyme';
import { initialState } from '../store/reducers';
import { App } from './App';

describe('App', () => {
  test('renders with books', () => {
    const props = {
      page: 1,
      itemsPerPage: 20,
      isLoading: false,
      error: null,
      books: [
        {
          id: 1,
          book_author: ['mocked'],
          book_title: 'mocked',
          book_publication_year: 2020,
          book_publication_country: 'mocked',
          book_publication_city: 'mocked',
          book_pages: 104
        }
      ],
      totalRecords: 0,
      fetchBooks: jest.fn().mockReturnValue()
    };

    const wrapper = shallow(<App {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('renders with no books error', () => {
    const props = {
      ...initialState,
      fetchBooks: jest.fn().mockReturnValue()
    };

    const wrapper = shallow(<App {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
