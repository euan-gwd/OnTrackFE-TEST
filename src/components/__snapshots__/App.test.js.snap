// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders with books 1`] = `
<div
  className="app-container"
>
  <h1
    className="app-header"
  >
    The Library
  </h1>
  <Connect(SearchBar) />
  <Connect(PageItemsFilter) />
  <ListGroup
    horizontal={null}
    variant={null}
  >
    <ListGroupItem
      active={false}
      disabled={false}
      key="1"
      variant={null}
    >
      <BookItem
        book={
          Object {
            "book_author": Array [
              "mocked",
            ],
            "book_pages": 104,
            "book_publication_city": "mocked",
            "book_publication_country": "mocked",
            "book_publication_year": 2020,
            "book_title": "mocked",
            "id": 1,
          }
        }
      />
    </ListGroupItem>
  </ListGroup>
  <div
    className="page-navbar"
  >
    <Connect(PageNavBar) />
  </div>
</div>
`;

exports[`App renders with no books error 1`] = `
<div
  className="app-container"
>
  <h1
    className="app-header"
  >
    The Library
  </h1>
  <Connect(SearchBar) />
  <Connect(PageItemsFilter) />
  <ListGroup
    horizontal={null}
    variant={null}
  >
    <Alert
      closeLabel="Close alert"
      show={true}
      transition={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "defaultProps": Object {
            "appear": false,
            "in": false,
            "mountOnEnter": false,
            "timeout": 300,
            "unmountOnExit": false,
          },
          "displayName": "Fade",
          "render": [Function],
        }
      }
      variant="warning"
    >
      Sorry, No Books found. Please try again.
    </Alert>
  </ListGroup>
</div>
`;
